src/main/java/bio/terra/service/snapshotbuilder/SnapshotRequestDao.java

```diff
@@ -0,0 +1,186 @@
+package bio.terra.service.snapshotbuilder;
+
+import bio.terra.common.DaoKeyHolder;
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.common.exception.NotFoundException;
+import bio.terra.model.SnapshotAccessRequest;
+import bio.terra.model.SnapshotAccessRequestResponse;
+import bio.terra.model.SnapshotBuilderRequest;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+@Repository
+public class SnapshotRequestDao {
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+  private static final SnapshotAccessRequestResponseMapper responseMapper =
+      new SnapshotAccessRequestResponseMapper();
+  private static final String requestIdField = "id";
+  private static final String datasetIdField = "dataset_id";
+  private static final String requestNameField = "request_name";
+  private static final String requestResearchPurposeField = "request_research_purpose";
+  private static final String snapshotBuilderRequestField = "snapshot_builder_request";
+  private static final String userEmailField = "user_email";
+  private static final String createDateField = "create_date";
+  private static final String updateDateField = "update_date";
+  private static final String statusField = "status";
+
+  @Autowired
+  public SnapshotRequestDao(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  private static class SnapshotAccessRequestResponseMapper
+      implements RowMapper<SnapshotAccessRequestResponse> {
+    public SnapshotAccessRequestResponse mapRow(ResultSet rs, int rowNum) throws SQLException {
+      return new SnapshotAccessRequestResponse()
```

Comment:
I'm not sure how much we need to worry about this now, or what the convention in TDR code is, but in general I wouldn't expect a DAO (or even a Service) to generate the model object used by a controller. The DAO should generate a DTO / model which would probably be different than the actual object returned by an API. If that's how it works now that's fine but in the future they could diverge.

The general idea is that we want to be able to easily change the API based on user need without worrying about needing to propagate the change through to the lower layers of code like DAO or Service.

Reply:
This makes sense to me and It brings up a different question for me. I recently had a UI PR that did mapping between UI types and the API types (https://github.com/DataBiosphere/terra-ui/pull/4486). Why wouldn't the API just return only what is needed in the UI and that 'mapping' would take place in the controller like you described here?